### PR TITLE
[flutter_tools] Include mode in app.start event, and forward app.start to DAP clients

### PR DESCRIPTION
**Pull request by DanTup**
This will allow DAP clients to know the final device ID and also the mode an app is running in (something it currently guesses, but may not be accurate).

Required to improve some messages in VS Code when trying to open DevTools when running in a mode that doesn't support it (https://github.com/Dart-Code/Dart-Code/issues/4273):

image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart [hunk 1]
```diff
@@ -382,7 +382,7 @@ class FlutterDebugAdapter extends FlutterBaseDebugAdapter {
   }
 
   /// Handles the app.start event from Flutter.
-  void _handleAppStart(Map<String, Object?> params) {
+  Future<void> _handleAppStart(Map<String, Object?> params) async {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart:385]: Why was this changed to be async?

Comment (DanTup) [reply packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart]: hmm, I'm not sure. Maybe I originally added some async code in here and didn't revert it properly (or maybe I was testing you 😄).

Removed, thanks!

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart]: Whew, glad I passed the test

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by DanTup** [comment]
@christopherfujino could you check what the Google Testing failure is?

**Comment by christopherfujino** [comment]
The check says to rebase...

**Comment by DanTup** [comment]
I think it just always says that now when the PR is out of date (which is almost always given how often master is commited to). Still, another both run might get lucky if it was an unrelated flake 😄

Thanks!

**Comment by christopherfujino** [comment]
> I think it just always says that now when the PR is out of date (which is almost always given how often master is commited to). Still, another both run might get lucky if it was an unrelated flake 😄

I mean the google testing check, I've definitely seen it say either pass or fail this week, even if the branch point is out of date.

**Comment by DanTup** [comment]
> I mean the google testing check

Oh, gotcha! I thought you meant Git (it seems to have a button for rebasing much more visibly than it used to), I'm not familiar with the Google Testing stuff.

Anyway, looks like it worked - thanks! :-)
